File: src/core/render.ts
/**
 * HTML helpers + overlay injection.
 *
 * The browser overlay (`overlay.js`) is loaded once at module init and injected
 * before `</body>`. This replaces the Worker's build-time inlining; the bytes
 * reaching the browser are identical.
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const here = dirname(fileURLToPath(import.meta.url));

/** Locate `overlay.js` next to the running module (dist/) or in src/ during dev. */
function loadOverlay(): string {
  for (const candidate of [join(here, 'overlay.js'), join(here, '..', 'overlay.js')]) {
    try {
      return readFileSync(candidate, 'utf8');
    } catch {
      // try next candidate
    }
  }
  throw new Error('overlay.js not found next to render module');
}

const OVERLAY_JS = loadOverlay();

/** The boot config injected as `window.__TDOC__` for the overlay. */
export interface OverlayConfig {
  slug: string;
  version: number;
  identity: OverlayIdentity | null;
  mode: 'published' | 'local' | 'fork';
  authConfigured: boolean;
  isOwner?: boolean;
  versions?: { n: number; created?: string | null }[];
  originalSlug?: string;
}

/** Minimal identity shape the overlay renders in its toolbar. */
export interface OverlayIdentity {
  login: string;
  avatar_url?: string | null;
  name?: string;
}

/** Build an {@link OverlayIdentity} from session-ish fields, honoring exactOptional. */
export function toOverlayIdentity(
  source: { login: string; avatar_url?: string | null; name?: string } | null,
): OverlayIdentity | null {
  if (!source) return null;
  const id: OverlayIdentity = { login: source.login };
  if (source.avatar_url != null) id.avatar_url = source.avatar_url;
  if (source.name != null) id.name = source.name;
  return id;
}

/** Escape `</script>` and HTML-comment openers so JSON can't break out of a `<script>`. */
export function safeJsonForScript(obj: unknown): string {
  return JSON.stringify(obj)
    .replace(/<\/script>/gi, '<\\/script>')
    .replace(/<!--/g, '<\\!--');
}

/** Full HTML escaping for interpolating untrusted strings into markup. */
export function escapeHtml(s: string | null | undefined): string {
  return (s ?? '').replace(
    /[&<>"']/g,
    (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' })[c] as string,
  );
}

/** Neutralize `--` so an untrusted string can't terminate an HTML comment. */
export function forHtmlComment(s: string | null | undefined): string {
  return (s ?? '').replace(/--/g, '-\\-');
}

/** Inject the overlay boot script + config before `</body>` (or append). */
export function injectOverlayCfg(rawHtml: string, cfg: OverlayConfig): string {
  const inject = `<script>window.__TDOC__ = ${safeJsonForScript(cfg)};</script>\n<script>${OVERLAY_JS}</script>`;
  return rawHtml.includes('</body>')
    ? rawHtml.replace('</body>', `${inject}\n</body>`)
    : rawHtml + inject;
}

File: internal/core/render.go
package core

import (
	"encoding/json"
	"strings"
)

// HTML helpers and overlay injection, ported from render.ts. The browser overlay
// is injected before </body>; the bytes reaching the browser are identical to the
// upstream worker's build-time inlining.

// OverlayIdentity is the minimal identity the overlay renders in its toolbar.
type OverlayIdentity struct {
	Login     string  `json:"login"`
	AvatarURL *string `json:"avatar_url,omitempty"`
	Name      string  `json:"name,omitempty"`
}

// OverlayConfig is the boot config injected as window.__ODOC__ for the overlay.
type OverlayConfig struct {
	Slug           string           `json:"slug"`
	Version        int              `json:"version"`
	Identity       *OverlayIdentity `json:"identity"`
	Mode           string           `json:"mode"`
	AuthConfigured bool             `json:"authConfigured"`
	IsOwner        bool             `json:"isOwner,omitempty"`
	Versions       []VersionRef     `json:"versions,omitempty"`
	OriginalSlug   string           `json:"originalSlug,omitempty"`
}

// VersionRef is one entry in the overlay's version picker.
type VersionRef struct {
	N       int     `json:"n"`
	Created *string `json:"created,omitempty"`
}

// SafeJSONForScript escapes </script> and HTML-comment openers so JSON can't
// break out of a <script> element. It uses an encoder with HTML escaping
// DISABLED so the byte output matches JavaScript's JSON.stringify (Go's default
// escapes <, >, & to \u00XX, which JS does not).
func SafeJSONForScript(v any) (string, error) {
	var buf strings.Builder
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		return "", err
	}
	s := strings.TrimRight(buf.String(), "\n") // Encoder appends a newline
	// Go's encoding/json always escapes U+2028/U+2029 (\u2028 / \u2029) even with
	// SetEscapeHTML(false); JavaScript's JSON.stringify emits them raw. Restore the
	// raw code points so the injected window.__ODOC__ bytes match upstream. (They
	// are valid inside a <script> JSON literal — only bare U+2028/9 in a JS string
	// literal would be a hazard, which this is not.)
	s = unescapeLineSep(s)
	s = strings.ReplaceAll(s, "</script>", `<\/script>`)
	s = strings.ReplaceAll(s, "<!--", `<\!--`)
	return s, nil
}

// unescapeLineSep rewrites genuine \u2028 / \u2029 JSON escape sequences (the
// 6-char ASCII text json emits) back to their raw code points, matching
// JavaScript JSON.stringify. A sequence is rewritten only when the backslash that
// starts it is itself unescaped (an even number of backslashes precede it), so a
// real escape is restored but the literal text \u2028 inside a string value
// (which json encodes as \\u2028 — an escaped backslash then u2028) is left
// intact. The preceding-backslash parity is tracked in the single forward pass.
func unescapeLineSep(s string) string {
	const esc8, esc9 = `\u2028`, `\u2029`
	if !strings.Contains(s, `\u202`) {
		return s
	}
	var b strings.Builder
	b.Grow(len(s))
	bs := 0 // consecutive backslashes immediately before the current index
	for i := 0; i < len(s); {
		if s[i] == '\\' && bs%2 == 0 {
			if strings.HasPrefix(s[i:], esc8) {
				b.WriteRune('\u2028')
				i += len(esc8)
				bs = 0
				continue
			}
			if strings.HasPrefix(s[i:], esc9) {
				b.WriteRune('\u2029')
				i += len(esc9)
				bs = 0
				continue
			}
		}
		if s[i] == '\\' {
			bs++
		} else {
			bs = 0
		}
		b.WriteByte(s[i])
		i++
	}
	return b.String()
}

var htmlEscaper = strings.NewReplacer(
	"&", "&amp;",
	"<", "&lt;",
	">", "&gt;",
	`"`, "&quot;",
	"'", "&#39;",
)

// EscapeHTML escapes a string for interpolation into markup.
func EscapeHTML(s string) string {
	return htmlEscaper.Replace(s)
}

// ForHTMLComment neutralizes -- so an untrusted string can't terminate an HTML comment.
func ForHTMLComment(s string) string {
	return strings.ReplaceAll(s, "--", `-\-`)
}

// InjectOverlayCfg injects the overlay boot script + config before </body>.
func InjectOverlayCfg(rawHTML, overlayJS string, cfg OverlayConfig) (string, error) {
	cfgJSON, err := SafeJSONForScript(cfg)
	if err != nil {
		return "", err
	}
	inject := "<script>window.__ODOC__ = " + cfgJSON + ";</script>\n<script>" + overlayJS + "</script>"
	if strings.Contains(rawHTML, "</body>") {
		return strings.Replace(rawHTML, "</body>", inject+"\n</body>", 1), nil
	}
	return rawHTML + inject, nil
}
